### PR TITLE
Added support for non-capturing lambda kernels.

### PR DESCRIPTION
**Pull request by MoFtZ**
Fixes https://github.com/m4rs-mt/ILGPU/issues/79 and https://github.com/m4rs-mt/ILGPU/issues/136.

Added heuristic for detecting non-capturing lambdas. The code generator then shifts all method arguments one place to the left, dropping the unused 'this' argument.

Also added a way for code generation on the background worker threads to record an exception, and return it to the original thread. This was needed because during unit tests, exceptions thrown in the worker thread terminate the unit test but do not mark the test as failed.

### CODE REVIEW

File: Src/ILGPU/Frontend/ILFrontend.cs [hunk 1]
```diff
@@ -160,20 +170,41 @@ private void DoWork()
...
+                try
+                {
+                    codeGenerationPhase.GenerateCodeInternal(
+                        current.Method,
+                        current.IsExternalRequest,
+                        detectedMethods,
+                        out Method method);
+                    current.SetResult(method);
+                }
+#pragma warning disable CA1031 // Do not catch general exception types
```
Comment (m4rs-mt) [inline Src/ILGPU/Frontend/ILFrontend.cs:182]: Maybe we should convert the pragma to a `[Suppress...]` attribute?

File: Src/ILGPU/Frontend/ILFrontend.cs [hunk 2]
```diff
@@ -160,20 +170,41 @@ private void DoWork()
...
 
-                    foreach (var detectedMethod in detectedMethods)
-                        processing.Push(new ProcessingEntry(detectedMethod, null));
+                    try
+                    {
+                        foreach (var detectedMethod in detectedMethods)
+                            processing.Push(new ProcessingEntry(detectedMethod, null));
+
+                    }
+#pragma warning disable CA1031 // Do not catch general exception types
```
Comment (m4rs-mt) [inline Src/ILGPU/Frontend/ILFrontend.cs:200]: nit: same here

File: Src/ILGPU/Frontend/CodeGenerator/CodeGenerator.cs [hunk 3]
```diff
@@ -82,17 +83,19 @@ private void SetupVariables()
...
                     case ILInstructionType.Ldarga:
                         variables.Add(new VariableRef(
-                            instruction.GetArgumentAs<int>(), VariableRefType.Argument));
+                            instruction.GetArgumentAs<int>() - LambdaArgumentOffset,
+                            VariableRefType.Argument));
                         break;
                     case ILInstructionType.Ldloca:
                         variables.Add(new VariableRef(
-                            instruction.GetArgumentAs<int>(), VariableRefType.Local));
+                            instruction.GetArgumentAs<int>() - LambdaArgumentOffset,
```
Comment (m4rs-mt) [inline Src/ILGPU/Frontend/CodeGenerator/CodeGenerator.cs:91]: Are you sure about this piece of code? The 'this' argument shouldn't change all local variable indexes in the function, should it?

File: Src/ILGPU/Util/MethodExtensions.cs [hunk 4]
```diff
@@ -27,6 +28,49 @@ static class MethodExtensions
...
+            // CompilerGenerated attribute because F# uses a different attribute
+            // i.e. [CompilationMapping(SourceConstructFlags.Closure)], which only exists
+            // in F#. As a side-effect, the detection therefore also allows instance
+            // methods on any class without instance fields or properties to be
+            // considered as a non-capturing lambda.
+            //
+            // IMPORTANT: It is possible for the lambda to capture static fields and
+            // properties, and still pass this detection because we do not inspect the
+            // IL instructions here. We are relying on the rest of the compilation to
+            // detect invalid cases.
```
Comment (m4rs-mt) [inline Src/ILGPU/Util/MethodExtensions.cs:58]: Makes sense to me 👍

File: Src/ILGPU/Frontend/CodeGenerator/CodeGenerator.cs (hunk 3 above)
Comment (MoFtZ) [reply Src/ILGPU/Frontend/CodeGenerator/CodeGenerator.cs:91]: Ah, my mistake, I misread that opcode thinking it was a variation to `Ldarg`. Will fix.

File: Src/ILGPU/Frontend/ILFrontend.cs (hunk 1 above)
Comment (MoFtZ) [reply Src/ILGPU/Frontend/ILFrontend.cs]: Done.

File: Src/ILGPU/Frontend/ILFrontend.cs (hunk 2 above)
Comment (MoFtZ) [reply Src/ILGPU/Frontend/ILFrontend.cs]: Done.

File: Src/ILGPU/Frontend/CodeGenerator/CodeGenerator.cs (hunk 3 above)
Comment (MoFtZ) [reply Src/ILGPU/Frontend/CodeGenerator/CodeGenerator.cs]: Done.

**Pull request merged**